### PR TITLE
Backports 0.15 pr26

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
this should be the final backporting pr for 0.15, there are likely a couple of 0.16 backports that should be included in dash 0.14.1 for stability / general bug fixing reasons

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
tests pass locally soo... ¯\\_(ツ)_/¯

File: src/validation.cpp [hunk 1]
```diff
@@ -4185,21 +4185,6 @@ bool LoadBlockIndex(const CChainParams& chainparams)
     return true;
 }
 
-static bool AddGenesisBlock(const CChainParams& chainparams, const CBlock& block, CValidationState& state)
```
Comment (UdjinM6) [inline src/validation.cpp]: this should be reverted, see previous comment

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -4766,11 +4766,6 @@ void CWallet::MarkReserveKeysAsUsed(int64_t keypool_id)
     }
 }
 
-bool CWallet::HasUnusedKeys(int min_keys) const
```
Comment (UdjinM6) [inline src/wallet/wallet.cpp]: that's a part of https://github.com/bitcoin/bitcoin/pull/11044

File: src/validation.cpp [hunk 3]
```diff
@@ -4198,24 +4225,21 @@ bool InitBlockIndex(const CChainParams& chainparams)
...
-                std::shared_ptr<const CBlock> shared_pblock = std::make_shared<const CBlock>(chainparams.DevNetGenesisBlock());
-                bool fCheckBlock = CheckBlock(*shared_pblock, state, chainparams.GetConsensus());
-                assert(fCheckBlock);
-                if (!AcceptBlock(shared_pblock, state, chainparams, nullptr, true, nullptr, nullptr))
-                    return false;
-            }
-        } catch (const std::runtime_error& e) {
-            return error("%s: failed to initialize block database: %s", __func__, e.what());
-        }
+    try {
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/validation.cpp]: do you mean remove the `if (!fReindex) {` that is currently there? I don't understand

[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/validation.cpp]: ah, yes, not _wrap_ but _unwrap_ 🙈 see original pr with `?w=1` https://github.com/bitcoin/bitcoin/pull/10758/files?w=1#diff-24efdb00bfbe56b140fb006b562cc70bL3886

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (PastaPastaPasta) [reply src/wallet/wallet.cpp]: see https://github.com/dashpay/dash/pull/3096/commits/32a3adfa4f23a65dcd16cee8bb8ebf49b84debbb and https://github.com/dashpay/dash/pull/3096/commits/6946848e027927522e2bf1cc4d0d55d07e6b580b

File: src/validation.cpp (hunk 1 above)
Comment (PastaPastaPasta) [reply src/validation.cpp]: see https://github.com/dashpay/dash/pull/3096/commits/29b5f382d6550f9a244fd1abe2d213a1c75cb1ff

File: src/validation.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/validation.cpp]: see https://github.com/dashpay/dash/pull/3096/commits/a7e200e9794c6c7cf86a5395ce72b178456e45f3

File: src/validation.cpp [hunk 4]
```diff
@@ -4172,18 +4200,17 @@ static bool AddGenesisBlock(const CChainParams& chainparams, const CBlock& block
...
+    // mapBlockIndex. Note that we can't use chainActive here, since it is
+    // set based on the coins db, not the block index db, which is the only
+    // thing loaded at this point.
+    if (mapBlockIndex.count(chainparams.GenesisBlock().GetHash()))
         return true;
 
-    // Use the provided setting for -txindex in the new database
-    fTxIndex = gArgs.GetBoolArg("-txindex", DEFAULT_TXINDEX);
-    pblocktree->WriteFlag("txindex", fTxIndex);
-
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/validation.cpp]: see https://github.com/dashpay/dash/pull/3096/commits/65ad8ad6a6f83ac75c31aa50c82d53c1fea4ba54

**Comment by PastaPastaPasta** [comment]
resolved all comments, tests passing locally, can squash some commits/remove the initial & reverts on utACK

**Comment by codablock** [comment]
Didn't do a full review yet, but I noticed that https://github.com/dashpay/dash/pull/3096/commits/6478562b3d4254c73aa973d24bc70fb4b42babbe is included. I backported the same PR, but based on #3101 to avoid the nasty merge conflicts. @Pasta what do you think about removing this specific backport from this PR and instead use my version (which I'll submit in a few minutes)?

**Comment by UdjinM6** [comment]
I like the #3101 way too (need to revert both 6478562 and a 8a216cc in this case). This should probably help with future backports too.

**Comment by UdjinM6** [comment]
Pls see https://github.com/UdjinM6/dash/commit/0f96089d969fe803e8cf89101f620d574d325c9a (note, this is after reverting 6478562 and 8a216cc, the later has an additional issue as it turned out).

**Comment by PastaPastaPasta** [comment]
dropped those two commits and cherry-picked Udjin's fix

**Comment by PastaPastaPasta** [comment]
dropped the original + revert commits

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -4333,21 +4333,18 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
...
                 fInternal = true;
             }
-            if (!setInternalKeyPool.empty()) {
-                nEnd = *(--setInternalKeyPool.end()) + 1;
-            }
-            if (!setExternalKeyPool.empty()) {
-                nEnd = std::max(nEnd, *(--setExternalKeyPool.end()) + 1);
-            }
-            // TODO: implement keypools for all accounts?
-            if (!walletdb.WritePool(nEnd, CKeyPool(GenerateNewKey(walletdb, 0, fInternal), fInternal))) {
```
Comment (codablock) [inline src/wallet/wallet.cpp]: I see you fixed the argument discrepancy in a later fix commit, but you omitted the TODO comment that you also removed by accident. We should either re-add that comment or figure out how to deal with multi-account derivation. My current assumption is that we can actually ignore multi-accounts and always use 0 (and thus change the GenerateNewKey method).

File: src/init.cpp [hunk 6]
```diff
@@ -1802,37 +1794,81 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
...
                 if (!ReplayBlocks(chainparams, pcoinsdbview)) {
                     strLoadError = _("Unable to replay blocks. You will need to rebuild the database using -reindex-chainstate.");
                     break;
                 }
+
+                // The on-disk coinsdb is now in a good state, create the cache
                 pcoinsTip = new CCoinsViewCache(pcoinscatcher);
-                LoadChainTip(chainparams);
 
                 deterministicMNManager->UpgradeDBIfNeeded();
```
Comment (codablock) [inline src/init.cpp]: `UpgradeDBIfNeeded` must happen after `LoadChainTip`, otherwise it becomes a no-op.

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -4346,6 +4428,8 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
                 setExternalKeyPool.insert(index);
             }
 
+            m_pool_key_to_index[pubkey.GetID()] = index;
+
```
Comment (codablock) [inline src/wallet/wallet.cpp]: NIT: this empty line was not added in the original PR

**Pull request merged**